File: go/collectors/gpu.go
package collectors

import (
	"os/exec"
	"strconv"
	"strings"

	"machinestate/models"
)

// CollectGPUInfo gathers GPU information
func CollectGPUInfo() models.GPUInfo {
	info := models.GPUInfo{
		GPUs: []models.GPUDevice{},
	}

	// Try NVIDIA first
	if gpus := getNvidiaGPUs(); len(gpus) > 0 {
		info.Available = true
		info.GPUs = gpus
		return info
	}

	// Try AMD
	if gpus := getAMDGPUs(); len(gpus) > 0 {
		info.Available = true
		info.GPUs = gpus
		return info
	}

	// Fallback to lspci detection
	if gpus := detectGPUsFromLspci(); len(gpus) > 0 {
		info.Available = true
		info.GPUs = gpus
		return info
	}

	return info
}

func getNvidiaGPUs() []models.GPUDevice {
	var gpus []models.GPUDevice

	// Check if nvidia-smi is available
	if _, err := exec.LookPath("nvidia-smi"); err != nil {
		return gpus
	}

	// Query GPU info
	cmd := exec.Command("nvidia-smi",
		"--query-gpu=index,name,temperature.gpu,utilization.gpu,memory.used,memory.total,power.draw",
		"--format=csv,noheader,nounits")
	output, err := cmd.Output()
	if err != nil {
		return gpus
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}

		fields := strings.Split(line, ", ")
		if len(fields) < 7 {
			continue
		}

		index, _ := strconv.Atoi(strings.TrimSpace(fields[0]))
		temp, _ := strconv.Atoi(strings.TrimSpace(fields[2]))
		util, _ := strconv.Atoi(strings.TrimSpace(fields[3]))
		memUsed, _ := strconv.ParseInt(strings.TrimSpace(fields[4]), 10, 64)
		memTotal, _ := strconv.ParseInt(strings.TrimSpace(fields[5]), 10, 64)
		power, _ := strconv.ParseFloat(strings.TrimSpace(fields[6]), 64)

		gpu := models.GPUDevice{
			Index:       index,
			Name:        strings.TrimSpace(fields[1]),
			Vendor:      "nvidia",
			Temperature: temp,
			Utilization: util,
			MemoryUsed:  memUsed * 1024 * 1024, // MiB to bytes
			MemoryTotal: memTotal * 1024 * 1024,
			PowerDraw:   power,
		}

		gpus = append(gpus, gpu)
	}

	return gpus
}

func getAMDGPUs() []models.GPUDevice {
	var gpus []models.GPUDevice

	// Check if rocm-smi is available
	if _, err := exec.LookPath("rocm-smi"); err != nil {
		return gpus
	}

	// Get temperature
	cmd := exec.Command("rocm-smi", "--showtemp", "--csv")
	tempOutput, err := cmd.Output()
	if err != nil {
		return gpus
	}

	// Parse temperature output
	tempMap := make(map[int]int)
	lines := strings.Split(string(tempOutput), "\n")
	for i, line := range lines {
		if i == 0 || line == "" {
			continue
		}
		fields := strings.Split(line, ",")
		if len(fields) >= 2 {
			idx, _ := strconv.Atoi(strings.TrimSpace(fields[0]))
			temp, _ := strconv.ParseFloat(strings.TrimSpace(fields[1]), 64)
			tempMap[idx] = int(temp)
		}
	}

	// Get utilization
	cmd = exec.Command("rocm-smi", "--showuse", "--csv")
	useOutput, err := cmd.Output()
	if err != nil {
		return gpus
	}

	useMap := make(map[int]int)
	lines = strings.Split(string(useOutput), "\n")
	for i, line := range lines {
		if i == 0 || line == "" {
			continue
		}
		fields := strings.Split(line, ",")
		if len(fields) >= 2 {
			idx, _ := strconv.Atoi(strings.TrimSpace(fields[0]))
			use, _ := strconv.ParseFloat(strings.TrimSpace(fields[1]), 64)
			useMap[idx] = int(use)
		}
	}

	// Get memory usage
	cmd = exec.Command("rocm-smi", "--showmemuse", "--csv")
	memOutput, err := cmd.Output()
	if err != nil {
		return gpus
	}

	lines = strings.Split(string(memOutput), "\n")
	for i, line := range lines {
		if i == 0 || line == "" {
			continue
		}
		fields := strings.Split(line, ",")
		if len(fields) >= 2 {
			idx, _ := strconv.Atoi(strings.TrimSpace(fields[0]))

			gpu := models.GPUDevice{
				Index:       idx,
				Name:        "AMD GPU " + strconv.Itoa(idx),
				Vendor:      "amd",
				Temperature: tempMap[idx],
				Utilization: useMap[idx],
			}

			gpus = append(gpus, gpu)
		}
	}

	return gpus
}

func detectGPUsFromLspci() []models.GPUDevice {
	var gpus []models.GPUDevice

	// Check if lspci is available
	if _, err := exec.LookPath("lspci"); err != nil {
		return gpus
	}

	cmd := exec.Command("lspci")
	output, err := cmd.Output()
	if err != nil {
		return gpus
	}

	lines := strings.Split(string(output), "\n")
	index := 0
	for _, line := range lines {
		lineLower := strings.ToLower(line)
		if strings.Contains(lineLower, "vga") || strings.Contains(lineLower, "3d") || strings.Contains(lineLower, "display") {
			// Extract GPU name
			parts := strings.SplitN(line, ":", 3)
			name := "Unknown GPU"
			if len(parts) >= 3 {
				name = strings.TrimSpace(parts[2])
			}

			vendor := "unknown"
			if strings.Contains(lineLower, "nvidia") {
				vendor = "nvidia"
			} else if strings.Contains(lineLower, "amd") || strings.Contains(lineLower, "radeon") {
				vendor = "amd"
			} else if strings.Contains(lineLower, "intel") {
				vendor = "intel"
			}

			gpu := models.GPUDevice{
				Index:  index,
				Name:   name,
				Vendor: vendor,
				// Temperature, utilization, memory not available via lspci
			}

			gpus = append(gpus, gpu)
			index++
		}
	}

	return gpus
}

File: zig/src/collectors/gpu.zig
const std = @import("std");
const utils = @import("../utils.zig");
const report = @import("../models/report.zig");

pub fn collectGPUInfo(allocator: std.mem.Allocator) ?report.GPUInfo {
    var devices = std.ArrayList(report.GPUDevice).empty;

    // Try NVIDIA first
    if (collectNvidiaGPU(allocator)) |gpu| {
        devices.append(allocator, gpu) catch {};
    }

    // Try AMD
    if (collectAMDGPU(allocator)) |gpu| {
        devices.append(allocator, gpu) catch {};
    }

    // Fallback to lspci for detection
    if (devices.items.len == 0) {
        collectLspciGPUs(allocator, &devices);
    }

    if (devices.items.len == 0) {
        return report.GPUInfo{
            .available = false,
        };
    }

    return report.GPUInfo{
        .available = true,
        .gpus = devices.toOwnedSlice(allocator) catch null,
    };
}

fn collectNvidiaGPU(allocator: std.mem.Allocator) ?report.GPUDevice {
    if (!utils.commandExists("nvidia-smi")) {
        return null;
    }

    const output = utils.runCommand(allocator, &[_][]const u8{
        "nvidia-smi",
        "--query-gpu=name,temperature.gpu,utilization.gpu,memory.used,memory.total,power.draw",
        "--format=csv,noheader,nounits",
    }) catch return null;

    const line = utils.trim(output);
    if (line.len == 0) return null;

    var parts = std.mem.splitSequence(u8, line, ", ");

    const name = parts.next() orelse return null;
    const temp_str = parts.next() orelse "0";
    const util_str = parts.next() orelse "0";
    const mem_used_str = parts.next() orelse "0";
    const mem_total_str = parts.next() orelse "0";
    const power_str = parts.next() orelse "0";

    return report.GPUDevice{
        .index = 0,
        .name = allocator.dupe(u8, name) catch return null,
        .vendor = "nvidia",
        .temperature = utils.parseInt(i32, temp_str, 0),
        .utilization = utils.parseInt(i32, util_str, 0),
        .memory_used = utils.parseInt(i64, mem_used_str, 0) * 1024 * 1024,
        .memory_total = utils.parseInt(i64, mem_total_str, 0) * 1024 * 1024,
        .power_draw = utils.parseFloat(power_str, 0),
    };
}

fn collectAMDGPU(allocator: std.mem.Allocator) ?report.GPUDevice {
    if (!utils.commandExists("rocm-smi")) {
        return null;
    }

    var temp: i32 = 0;
    var util: i32 = 0;

    const temp_output = utils.runCommand(allocator, &[_][]const u8{
        "rocm-smi", "--showtemp",
    }) catch "";

    if (temp_output.len > 0) {
        var lines = std.mem.splitScalar(u8, temp_output, '\n');
        while (lines.next()) |line| {
            if (std.mem.indexOf(u8, line, "Temperature")) |_| {
                for (line, 0..) |c, i| {
                    if (c >= '0' and c <= '9') {
                        var end = i;
                        while (end < line.len and (line[end] >= '0' and line[end] <= '9')) : (end += 1) {}
                        temp = utils.parseInt(i32, line[i..end], 0);
                        break;
                    }
                }
            }
        }
    }

    const util_output = utils.runCommand(allocator, &[_][]const u8{
        "rocm-smi", "--showuse",
    }) catch "";

    if (util_output.len > 0) {
        var lines = std.mem.splitScalar(u8, util_output, '\n');
        while (lines.next()) |line| {
            if (std.mem.indexOf(u8, line, "GPU use")) |_| {
                for (line, 0..) |c, i| {
                    if (c >= '0' and c <= '9') {
                        var end = i;
                        while (end < line.len and (line[end] >= '0' and line[end] <= '9')) : (end += 1) {}
                        util = utils.parseInt(i32, line[i..end], 0);
                        break;
                    }
                }
            }
        }
    }

    if (temp == 0 and util == 0) return null;

    return report.GPUDevice{
        .index = 0,
        .name = allocator.dupe(u8, "AMD GPU") catch "AMD GPU",
        .vendor = "amd",
        .temperature = temp,
        .utilization = util,
    };
}

fn collectLspciGPUs(allocator: std.mem.Allocator, devices: *std.ArrayList(report.GPUDevice)) void {
    if (!utils.commandExists("lspci")) return;

    const output = utils.runCommand(allocator, &[_][]const u8{
        "lspci",
    }) catch return;

    var lines = std.mem.splitScalar(u8, output, '\n');
    var idx: i32 = 0;

    while (lines.next()) |line| {
        const lower = line;
        if (std.mem.indexOf(u8, lower, "VGA") != null or
            std.mem.indexOf(u8, lower, "3D") != null or
            std.mem.indexOf(u8, lower, "Display") != null)
        {
            // Format: "02:00.0 VGA compatible controller: Intel Corporation..."
            // Split by ":" and take third part
            var colon_count: usize = 0;
            var name_start: usize = 0;
            for (line, 0..) |c, i| {
                if (c == ':') {
                    colon_count += 1;
                    if (colon_count == 2) {
                        name_start = i + 1;
                        break;
                    }
                }
            }

            if (name_start == 0 or name_start >= line.len) continue;
            const name = utils.trim(line[name_start..]);
            if (name.len == 0) continue;

            const vendor: []const u8 = if (std.mem.indexOf(u8, lower, "nvidia") != null or std.mem.indexOf(u8, lower, "NVIDIA") != null)
                "nvidia"
            else if (std.mem.indexOf(u8, lower, "amd") != null or std.mem.indexOf(u8, lower, "AMD") != null or std.mem.indexOf(u8, lower, "radeon") != null)
                "amd"
            else if (std.mem.indexOf(u8, lower, "intel") != null or std.mem.indexOf(u8, lower, "Intel") != null)
                "intel"
            else
                "unknown";

            devices.append(allocator, report.GPUDevice{
                .index = idx,
                .name = allocator.dupe(u8, name) catch continue,
                .vendor = vendor,
            }) catch continue;

            idx += 1;
        }
    }
}
